### PR TITLE
 lib/repo: Don't syncfs or fsync() dirs if fsync opt is disabled

### PR DESCRIPTION
**Pull request by cgwalters**
There are use cases for not syncing at all; think build cache repos, etc. Let's
be consistent here and make sure if fsync is disabled we do no sync at all.

I chose this opportunity to add tests using the shiny new strace fault
injection.  I can forsee using this for a lot more things, so I made
the support for detecting things generic.

Related: #1184

### CODE REVIEW

**Comment by cgwalters** [comment]
> strace: ptrace(PTRACE_TRACEME, ...): Operation not permitted

Ah, right 😦

File: tests/basic-test.sh [hunk 1]
```diff
@@ -809,8 +809,11 @@ cd ${test_tmpdir}
 rm -rf test2-checkout
 mkdir -p test2-checkout
 cd test2-checkout
-touch should-not-be-fsynced
-$OSTREE commit ${COMMIT_ARGS} -b test2 -s "Unfsynced commit" --fsync=false
+echo 'should not be fsynced' > should-not-be-fsynced
+if ! skip_one_without_strace_fault_injection; then
+    ${CMD_PREFIX} strace -o /dev/null -f -e inject=syncfs,fsync,sync:error=EPERM ostree --repo=${test_tmpdir}/repo commit ${COMMIT_ARGS} -b test2 -s "Unfsynced commit" --fsync=false
+    echo "ok fsync disabled"
+fi
```
Comment (jlebon) [inline tests/basic-test.sh]: Maybe another test here for the opposite? I.e. checking that without `--fsync=false`, we fail here with `EPERM`.

File: tests/libtest.sh [hunk 2]
```diff
@@ -540,6 +540,29 @@ skip_without_user_xattrs () {
     fi
 }
 
+# https://brokenpi.pe/tools/strace-fault-injection
+_have_strace_fault_injection=''
+have_strace_fault_injection() {
+    if test "${_have_strace_fault_injection}" = ''; then
+        if strace -P ${test_srcdir}/libtest-core.sh -e inject=read:retval=0 cat ${test_srcdir}/libtest-core.sh >out.txt &&
```
Comment (jlebon) [inline tests/libtest.sh]: This is pretty cool!

File: tests/libtest.sh [hunk 3]
```diff
@@ -540,6 +540,29 @@ skip_without_user_xattrs () {
...
+           test '!' -s out.txt; then
+            _have_strace_fault_injection=0
```
Comment (jlebon) [inline tests/libtest.sh]: I think I would change this to the less confusing `=1` and `=0` in the `else` branch, and then make the last line of the function `test "${_have_strace_fault_injection}" = '1'`.

File: tests/libtest.sh [hunk 4]
```diff
@@ -512,9 +512,21 @@ os_repository_new_commit ()
 }
 
 # Usage: if ! skip_one_without_user_xattrs; then ... more tests ...; fi
+_have_user_xattrs=''
+have_user_xattrs() {
+    if test "${_have_user_xattrs}" = ''; then
+        touch test-xattrs
+        if setfattr -n user.testvalue -v somevalue test-xattrs 2>/dev/null; then
+            _have_user_xattrs=0
```
Comment (jlebon) [inline tests/libtest.sh]: And similarly here too (re. swapping `=1` and `=0`). It's just much easier to reason about I find.

**Comment by cgwalters** [comment]
Fixups ⬆️

**Comment by jlebon** [comment]
@rh-atomic-bot r+ f524af7

**Comment by jlebon** [comment]
@rh-atomic-bot retry

**Pull request closed without merging**